File: DataStructures/Arrays/Contain Common Items.py
'''
Given 2 arrays, create a function that let's a user know (true/false) whether these two arrays contain any common items, 
e.g.
array1 = ['a', 'b', 'c', 'x']
array2 = ['z', 'y', 'i']
should return false.

e.g.
array1 = ['a', 'b', 'c', 'x']
array2 = ['z', 'y', 'x'];
should return true.

2 parameters - arrays - no size limit and return values should be True or False
'''


array1 = ['a', 'b', 'c', 'x'];
array2 = ['z', 'y', 'a'];


# brute force
# or more accurately Time Complexity O(a*b) and Space Complexity O(1)
def containsCommonItem(arr1, arr2):
    len1 = len(arr1)
    len2 = len(arr2)

    # in case no elements in arr1 and/or arr2
    if len1 == 0 or len2 == 0 or (len1 == 0 and len2 == 0):
        return False

    # in case no empty arrays
    for i in range(len2):
        for j in range(len2):
            if arr1[i] == arr2[j]:
                return True
    
    return False


# Better approach
# O(a + b) Time Complexity and Space Complexity O(a)
def containsCommonItem2(arr1, arr2):
    len1 = len(arr1)
    len2 = len(arr2)

    # empty array(s)
    if len1 == 0 or len2 == 0 or (len1 == 0 and len2 == 0):
        return False

    # nonempty arrays
    dict1 = {arr1[i]: "{arr1[i]}" for i in arr1}
    
    for j in range(len(arr2)):
        if arr2[j] in dict1:
            return True
    
    return False

File: Algos/Contain Common Items.js
// Given 2 arrays, create a function that let's a user know (true/false) whether these two arrays contain any common items
// For Example:
// const array1 = ['a', 'b', 'c', 'x'];//const array2 = ['z', 'y', 'i'];
// should return false.
//-----------
// const array1 = ['a', 'b', 'c', 'x'];//const array2 = ['z', 'y', 'x'];
// should return true.

// 2 parameters - arrays - no size limit
// return true or false


const array1 = ['a', 'b', 'c', 'x'];
const array2 = ['z', 'y', 'a'];


// brute force O(a*b) Time - Space O(1)
function containsCommonItem(arr1, arr2) {
  // empty array(s)
  if (!arr1.length || !arr2.length || (!arr1.length && !arr2.length)) {
    return false;
}
  // nonempty arrays
  for (let i=0; i < arr1.length; i++) {
    for ( let j=0; j < arr2.length; j++) {
      if(arr1[i] === arr2[j]) {
        return true;
      }
    }
  }
  return false;
}


// Better approach Time O(a + b) - Space O(a)
function containsCommonItem2(arr1, arr2) {
   // empty array(s)
   if (!arr1.length || !arr2.length || (!arr1.length && !arr2.length)) {
    return false;
  }
  // nonempty arrays
  let map = new Map();
  // create a hash table with keys being the nums in array1
  for (let i=0; i < arr1.length; i++) {
    if(!map[arr1[i]]) {
      const item = arr1[i];
      map[item] = true;
    }
  }
  // loop through array2 and check if item in exists in the hash table. 
  for (let j=0; j < arr2.length; j++) {
    if (map[arr2[j]]) {
      return true;
    }
  }
  return false;
}


// built-in method
function containsCommonItem3(arr1, arr2) {
  return arr1.some(item => arr2.includes(item))
}


console.log(
containsCommonItem(array1, array2),
containsCommonItem2(array1, array2),
containsCommonItem3(array1, array2)
)





// The some() method in JavaScript:
// checks if any array elements pass a test(provided as a callback function).
// executes the callback function once for each array element.
// returns true (and stops) if the function returns true for one of the array elements.
// returns false if the function returns false for all of the array elements.
// does not execute the function for empty array elements.
// does not change the original array.


// -----------------------------------------

// MAP vs Objects in JavaScript
// Map is a collection of keyed data items, just like an Object, which only allows string keys). But the main difference is that Map allows keys of any type.

// Methods and properties are:

// new Map() – creates the map.
// map.set(key, value) – stores the value by the key.
// map.get(key) – returns the value by the key, undefined if key doesn’t exist in map.
// map.has(key) – returns true if the key exists, false otherwise.
// map.delete(key) – removes the value by the key.
// map.clear() – removes everything from the map.
// map.size – returns the current element count.